pyro/infer/predictive.py

```diff
@@ -219,3 +220,71 @@ def get_vectorized_trace(self, *args, **kwargs):
                                             parallel=self.parallel, model_args=args, model_kwargs=kwargs)
         return _predictive(self.model, posterior_samples, self.num_samples,
                            return_trace=True, model_args=args, model_kwargs=kwargs)
+
+
+def log_likelihood(
+    model: Callable,
+    posterior_samples: Dict[str, torch.Tensor] = None,
+    guide: Callable = None,
+    num_samples: int = None,
+    parallel: bool = True,
+) -> Callable:
+    """Calculates the log likelihood of the data for a model with observed
+    sites, given posterior samples (or a guide that can generate them).
+
+    :param model: Model function containing pyro primitives. Must have observed
+        sites conditioned on data.
+    :type model: Callable
+    :param posterior_samples: Dictionary of posterior samples, defaults to None
+    :type posterior_samples: Dict[str, torch.Tensor], optional
+    :param guide: Guide function, defaults to None
+    :type guide: Callable, optional
+    :param num_samples: Number of posterior samples to evaluate, defaults to None
+    :type num_samples: int, optional
+    :param parallel: Whether to use vectorization, defaults to True
+    :type parallel: bool, optional
+    :rtype: Callable
+    """
+
+    # define the function to return the likelihood values
+    def log_like_helper(*args, **kwargs) -> torch.Tensor:
+        # trace once to get observed sites
+        trace = pyro.poutine.trace(model).get_trace(*args, **kwargs)
+        observed = {
+            name: site["value"]
+            for name, site in trace.nodes.items()
+            if site["type"] == "sample" and site["is_observed"]
+        }
+        # now trace it and extract the likelihood from observed sites
+        predictive = Predictive(
+            model, posterior_samples, guide, num_samples, (), parallel)
+        # use vectorized trace
+        if parallel:
+            log_like = dict()
+            trace = predictive.get_vectorized_trace(*args, **kwargs)
+            for obs_name, obs_val in observed.items():
+                obs_site = trace.nodes[obs_name]
+                log_like[obs_name] = obs_site["fn"].log_prob(obs_val).detach().cpu()
+        # iterate over samples from posterior if model can't be vectorized
+        else:
+            all_samples = {
+                k: v for k, v in predictive(*args, **kwargs).items()
+                if k not in observed}
```

Comment:
@ecotner Sorry for late response!

I am a bit worried that the implementation is tracing the model 2 times (1 to get samples, 1 to get the trace of each sample). But this seems pretty clean to me. Please feel free to add a TODO here or if you want to address it, you can add an argument in `_predictive_sequential`, e.g. `collect_fn=None`, so that if `collect_fn` is not None, `_predictive` will return the collection of values obtained by `collect_fn`. For example, to get log_prob at observed sites, in `log_likelihood`, we can use
```python
collect_fn = lambda msg: msg["fn"].log_prob(msg["value"])
```
and replace line 43 of this file by
```python
collected.append({site: collect_fn(trace.nodes[site]) for site in return_site_shapes})
```

Reply:
ooh that's a great idea!